### ISSUE TITLE
This GitHub source cannot be replicated. why??

### ISSUE THREAD

**Issue by peternara** [bassyess/CFCD#7]
I am grateful for your source @bassyess disclosure. 

I am replicating the 2nd stage from this source. However, unlike the experimental results in the paper, I couldn't achieve that outcome. It's not even close.

The parts I changed are very minor for execution. I did not alter the important elements like the network, sampling, or loss. 

First, for the experiment, I had to create a landmark_scores.txt file, and I used infer_epoch() located inside /core/trainer.py.
Below is a sample of the results for this.
```
/mnt/d/benchmarksets/GLDV2/train/e/2/f/e2f5af74f548acdf.jpg 33798 33798 1.0
/mnt/d/benchmarksets/GLDV2/train/5/f/d/5fd3f99be4e1b4c3.jpg 2334 2334 0.9999879598617554
/mnt/d/benchmarksets/GLDV2/train/b/b/9/bb9af5883b2bcc86.jpg 24263 24263 1.0
/mnt/d/benchmarksets/GLDV2/train/b/b/0/bb0a05119cf31f48.jpg 13294 13294 1.0
/mnt/d/benchmarksets/GLDV2/train/c/6/1/c61e44c807870e93.jpg 65612 65612 1.0
/mnt/d/benchmarksets/GLDV2/train/8/8/1/881965fb627215cd.jpg 45869 45869 1.0
/mnt/d/benchmarksets/GLDV2/train/4/1/2/4122c83d5a39e7ef.jpg 56906 56906 0.9999991655349731
/mnt/d/benchmarksets/GLDV2/train/4/9/7/4974a9d1d6b6e8f9.jpg 3744 3744 0.9964727759361267
/mnt/d/benchmarksets/GLDV2/train/4/0/8/40878d511abcf02a.jpg 76724 76724 1.0
/mnt/d/benchmarksets/GLDV2/train/7/6/f/76f57ea17931b49a.jpg 57688 57688 0.9998284578323364
/mnt/d/benchmarksets/GLDV2/train/a/1/9/a19f3d8fdfb32078.jpg 40023 40023 1.0
/mnt/d/benchmarksets/GLDV2/train/7/d/6/7d6754cbfd505a4b.jpg 9567 9567 1.0
/mnt/d/benchmarksets/GLDV2/train/1/6/d/16d0b8a388d5d76a.jpg 34543 34543 0.9999994039535522
/mnt/d/benchmarksets/GLDV2/train/0/f/8/0f8c50f7ed984770.jpg 47496 47496 1.0
/mnt/d/benchmarksets/GLDV2/train/e/6/6/e6623dcc31749f38.jpg 15219 15219 1.0
/mnt/d/benchmarksets/GLDV2/train/8/9/6/896e39dd9016600f.jpg 78202 78202 1.0
/mnt/d/benchmarksets/GLDV2/train/3/7/f/37f1df1cccdaaf5d.jpg 53312 53312 0.9978245496749878
/mnt/d/benchmarksets/GLDV2/train/a/5/7/a57586cd519eac1b.jpg 5384 5384 0.9999994039535522
/mnt/d/benchmarksets/GLDV2/train/2/c/c/2ccf9f75b74c5c06.jpg 37794 37794 1.0
/mnt/d/benchmarksets/GLDV2/train/7/e/6/7e614d6c3deb7188.jpg 63837 63837 1.0
/mnt/d/benchmarksets/GLDV2/train/d/4/1/d41e28dc62254eff.jpg 1694 1694 0.9999998807907104
/mnt/d/benchmarksets/GLDV2/train/7/8/5/785da6c757c6cdf0.jpg 75724 75724 0.9999943971633911
/mnt/d/benchmarksets/GLDV2/train/5/5/1/55110ec3a0ad2152.jpg 4665 4665 1.0
/mnt/d/benchmarksets/GLDV2/train/4/4/3/443f88def82ab1b6.jpg 61305 61305 0.9999998807907104
/mnt/d/benchmarksets/GLDV2/train/5/d/0/5d0b41e179f1a71e.jpg 19094 19094 1.0
/mnt/d/benchmarksets/GLDV2/train/5/1/6/5166e335c570d477.jpg 50874 50874 0.9999957084655762
...
```

# 1-stage 
I experimented with ResNet101, and for this, I used the resnet101_cfcd_s1_8gpu.yaml file.
The base lr in this file is set to BASE_LR: 0.005, but the paper has it set to 0.01, so I changed it to 0.01.
0.005 is the lr for the 2-stage.

![image](https://github.com/user-attachments/assets/62ff7b3c-b0fc-4bd8-9166-47f58ee25b7c)

After that, I proceeded with the training.
However, I was not satisfied with the training results. The performance was significantly lower than the 1-stage model provided by the author.

my result
![image](https://github.com/user-attachments/assets/35bc93ad-5a71-4e81-a372-471927610724)

 
author result
![image](https://github.com/user-attachments/assets/bb90921a-e2b9-429a-b2b3-00904fcebf00)

It appears that overfitting is occurring in the training loss.
![image](https://github.com/user-attachments/assets/013752f2-925d-479e-9964-a4b6380fb06d)

The strange thing is that even though the loss is increasing, the top-1 error is decreasing, and the top-5 error seems to be slightly increasing.
When applying Arcface loss, performance may drop, but there is no issue with the phenomenon shown in this training graph. This is also strange. Is it a characteristic of Madacos?  I can't understand why.

# 2-stage

The even stranger part is the 2-stage.  I didn't change anything in the 2-stage. Since my 1-stage model produced low results, I used the model provided by the author. The results are disastrous.

my result
![image](https://github.com/user-attachments/assets/4cbe501c-63b5-4e4d-b32c-6aecf79f057f)

author result
![image](https://github.com/user-attachments/assets/fe6a3215-d115-42e2-b25c-99fc707b83c7)

The training-related graph is also strange. The error is increasing.
![image](https://github.com/user-attachments/assets/2343a81b-d6b8-48e0-96a7-9fbec71e9268)

@bassyess, can you explain this phenomenon? Once again, I used your provided source/model exactly as it is.

Could it be that others were able to replicate it differently than I did?

**Comment by Zy-Zhang** [comment]
Hi @peternara , I'm not one of the authors but I also have interest in this work. I might could provide some explanation on the 1st stage because I've found a similar phenomena when I simply applying MadaCos to my model. As you can see, the training loss starts to increase but the Top1/5 error decreases slightly. This is normal since the scale (s) and margin (m) keep increasing with the training processes so the loss function changes as well (as shown in Figure 6 ). You could regard this work (at least the 1st stage) as a curriculum learning that keeps making the training objective harder and harder before the whole training become stable. If you try to manually stop increasing the scale and margin at a specific training epoch/iteration, you will observe that the loss starts to decrease latter since the training objective stop becoming more difficult. As to the 2nd stage, I have no idea since I think the triplet loss has marginal impacts on the final performance (as shown in Table 2). Hope this comment could help.
